src/Components/Server/src/Circuits/CircuitHost.cs

```diff
@@ -77,29 +78,36 @@ public static void SetCurrentCircuitHost(CircuitHost circuitHost)
 
         public Circuit Circuit { get; }
 
-        public IClientProxy Client { get; }
+        public IClientProxy Client { get; set; }
 
         public IJSRuntime JSRuntime { get; }
 
         public RemoteRenderer Renderer { get; }
 
         public RendererRegistry RendererRegistry { get; }
-
+        public IList<ComponentDescriptor> Descriptors { get; }
         public IServiceProvider Services { get; }
 
+        public Task<IEnumerable<string>> PrerrenderComponentAsync(Type componentType, ParameterCollection parameters)
```

Comment:
I still don't love that this is `IEnumerable<string>` but it seem OK for now.

Reply:
@rynowak For the MVC case we need to buffer the output as we need to write it synchronously. We could buffer it on a byte array, but I believe that would be more painful.
Another option would be to abstract it behind a type and provide overloads to write it to a stream/pipe (essentially an IHtmlContentResult like type).